ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveRelFieldTrimmer.java

```diff
@@ -674,10 +677,14 @@ public TrimResult trimFields(Project project, ImmutableBitSet fieldsUsed,
     // set columnAccessInfo for ViewColumnAuthorization
     for (Ord<RexNode> ord : Ord.zip(project.getProjects())) {
       if (fieldsUsed.get(ord.i)) {
-        if (this.columnAccessInfo != null && this.viewProjectToTableSchema != null
-            && this.viewProjectToTableSchema.containsKey(project)) {
+        if (this.viewProjectToTableSchema != null && this.viewProjectToTableSchema.containsKey(project)) {
           Table tab = this.viewProjectToTableSchema.get(project);
-          this.columnAccessInfo.add(tab.getCompleteName(), tab.getAllCols().get(ord.i).getName());
+          if (this.directColumnAccessInfo != null) {
+            this.directColumnAccessInfo.add(tab.getCompleteName(), tab.getAllCols().get(ord.i).getName());
+          }
+          if (this.allColumnAccessInfo != null) {
+            this.allColumnAccessInfo.add(tab.getCompleteName(), tab.getAllCols().get(ord.i).getName());
+          }
```

Comment:
this is actually duplicates the `ColumnAccessInfo` at a lot of places; wouldn't it make sense to extend `ColumnAccessInfo` internally; and enable it to "mark" columns; and add a method which is able to retrieve columns based on usage or not.

right now `CAI` is a `Map< TableNameString, Set<ColumnNameString>>`; how about changing this to a `Map<TableNameString, Set<ColumnAccess>` where `ColumnAccess` would be the `ColumnNameString` and some extra info which is added here.

Reply:
or...I'm right now thinking about deciding whether to use the "usage aware" one or not beforehand - and use a specifically tailored `ColumnAccesInfo` implementation

Reply:
I have used a ColumnAccess internal object to represent this. Now it is cleaner because it uses a single ds. However, I kept the current APIs in ColumnAccessInfo so I do not have to make any changes all over for the time being. Can you take another look? Thanks

Reply:
Cool!